### PR TITLE
Refactored workload API token management for better security and implemented generic API token dispenser

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
This PR heavily reworks the current mechanism of generating API tokens used by pipeline workloads to authenticate to the ZenML server API to implement additional scoping and security validation checks for generated workload API tokens and to remove the previous time-based limitations. Here are some highlights:

* API keys are no longer exposed to the pipeline workload containers
* workload API tokens are no longer time-restricted and can be used to run long-running pipelines without having to reconfigure the default time expiration.
* workload API tokens can be scoped to a schedule, a pipeline run or a pipeline step run (useful for step operators). The validity of these tokens is tied to the lifecycle of the object(s) to which it is scoped:
  * schedule scoped API tokens can only be used to authenticate for as long as the schedule itself exists and is active
  * pipeline run scoped API tokens can only be used to authenticate for as long as the pipeline run exists and the pipeline run hasn't concluded (i.e. the pipeline run status is not FAILED or COMPLETED)
  * step run scoped API tokens can only be used to authenticated for as long as the pipeline run step exists and the step run hasn't concluded (i.e. the step run status is not FAILED or COMPLETED).
* the API key scope and authorized device scope are transferred from the API token used to authorize the client to the generated workload API tokens, just as before

In addition to this, the current restrictions concerning which types of API tokens can be used to generate workload API tokens have been removed. All types of accounts and API tokens (authorized devices, service accounts, dashboard session tokens) can be used to generate API tokens. Additional workload API tokens can be generated by pipeline workloads authenticated with workload API tokens and the scoping rules are applied incrementally:

* a schedule scoped workload API token can be used to generate workload API tokens for the same schedule and any pipeline run and step run
* a pipeline run scoped workload API token can be used to generate workload API tokens for the same pipeline run and any step run
* a step run scoped workload API token can only be used to generate workload API tokens for the same step run 

These improvements add a layer of security to workload API tokens that was previously only implemented via time-based expiration that was difficult to work with in the context of long-running pipelines:

* API tokens scoped for a schedule are only valid while the schedule is configured and active. If accidentally leaked, they can be invalidated by deactivating or deleting the schedule.
* API tokens scoped for a pipeline run are only valid while the pipeline is running. Once the pipeline is completed or failed, they are no longer valid.
* API tokens scoped for a step run are only valid while the step is running. Once the step is completed or failed, they are no longer valid.
* in addition to the above, all generated workload tokens inherit the API key scope or authorized device scope from the current authorization context. Therefore deleting, rotating or disabling an API key or authorized device automatically invalidates all workload API tokens generated while authenticated with that API key or authorized device.

The existing api_token endpoint can now also be used to issue generic API tokens than expire after one hour and can be used to make programmatic calls to the API for automation purposes.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

## Summary by CodeRabbit

- **New Features**
	- Introduced a new attribute for managing the lifetime of generic API tokens.
	- Enhanced OAuth login process with additional parameters for token generation.
	- Added a logging mechanism to warn about non-expiring API tokens.

- **Bug Fixes**
	- Improved security by ensuring appropriate token management based on authentication methods.

- **Documentation**
	- Updated documentation to include details about the new API token lifetime attribute.

### CODE REVIEW

**Comment by htahir1** [comment]
@coderabbitai review

File: src/zenml/zen_server/routers/auth_endpoints.py
Comment (schustmi) [inline src/zenml/zen_server/routers/auth_endpoints.py:537]: I don't understand this early return here: Why is the verifying whether the user has permissions to create a pipeline run after this check? Is it not possible to request a pipeline scoped token without a device or API key?

My understanding would be that we want to maybe do this permission check if `pipeline_id` or `schedule_id` are set? Even that's not perfect in case of an unlisted pipeline, but covers more cases than the current implementation?

File: src/zenml/zen_server/routers/auth_endpoints.py
Comment (schustmi) [reply src/zenml/zen_server/routers/auth_endpoints.py:201]: This seems to be missing indeed

File: src/zenml/config/server_config.py
Comment (schustmi) [inline src/zenml/config/server_config.py:263]: We can use `PositiveInt` from `pydantic` to do this validation for us automatically (or `NonNegativeInt` in case we want to allow 0 seconds?)

**Review by schustmi — CHANGES_REQUESTED** [review]

**Comment by htahir1** [comment]
@schustmi Small note: If we can get this PR over the line and give to aagya we might be able to do it before she leaves... not sure but just wanted to add that if you can "take this over" so to speak and its an easy win, id do that

**Comment by schustmi** [comment]
Yeah I wanted to, but really unsure about the early return comment, I think we should probably wait for Stefan's input on this

File: src/zenml/orchestrators/utils.py
Comment (schustmi) [inline src/zenml/orchestrators/utils.py:118]: @stefannica maybe I'm missing something, but we show this warning message even if the user is currently authenticated with device/SA, no?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/zen_server/routers/auth_endpoints.py
Comment (stefannica) [reply src/zenml/zen_server/routers/auth_endpoints.py:561]: @schustmi , this endpoint is to be used in two distinct situations:

* from the dashboard, to generate 1 hour API tokens and display them to users (which is why we return early here)
* from the CLI, to generate these so called "workload" API tokens used by pipeline steps to connect to ZenML

The distinction is made by how the user is authenticated: if it's authenticated with an API key or through the web login (authorized device) flow, they are issued a "workload" API token (for which either a pipeline ID or schedule ID or both should theoretically be set, but I'll leave it to you to decide if we can also enforce that rule). If it's not an API key nor a device, it's a dashboard API token, in which case we issue a generic 1 hour token.

You can totally take this over and do whatever you like with this endpoint if you have better ideas for how we should make this available to users. My only concern is this: API tokens that are not tied to an API key nor an authorized device should be short lived, because they cannot be revoked in case of security incidents unless the entire user account is deactivated or deleted.

File: src/zenml/orchestrators/utils.py
Comment (stefannica) [reply src/zenml/orchestrators/utils.py]: Not after merging in my other PR (zenml login).

**Comment by stefannica** [comment]
@coderabbitai review

File: src/zenml/zen_server/auth.py
Comment (schustmi) [inline src/zenml/zen_server/auth.py:410]: If I get this right, every time such a workload token is used to send a request to the server, this code is running, correct? Do you think this will cause any trouble at scale as this is running up to 3 additional database queries per request?

**Review by schustmi — COMMENTED** [review]
Love the way this turned out, just one small comment from my side

File: src/zenml/zen_server/auth.py
Comment (stefannica) [reply src/zenml/zen_server/auth.py:410]: I guess I could implement a local time-based cache that would delay re-fetching the same object from the DB for up to 30 seconds. We should keep this contained to a limited number of entries though, otherwise the memory requirements will explode during high loads.

Comment (schustmi) [reply src/zenml/zen_server/auth.py:410]: Potentially yeah, but not necessary for now I hope. We're considerably reducing the amounts of server requests during pipeline execution in this and the next release, which means this hopefully doesn't matter as much

**Review by schustmi — APPROVED** [review]

File: src/zenml/zen_server/auth.py
Comment (stefannica) [reply src/zenml/zen_server/auth.py:475]: @schustmi I added a simple in-memory cache for objects that we want to cache in the ZenML server

File: src/zenml/zen_server/cache.py
Comment (schustmi) [inline src/zenml/zen_server/cache.py:65]: I don't see the `get_or_cache` method anywhere

File: src/zenml/zen_server/cache.py
Comment (schustmi) [inline src/zenml/zen_server/cache.py:134]: Is it possible to annotate `self.cache: OrderedDict[...]` to get rid of this type ignore?

File: src/zenml/zen_server/cache.py
Comment (stefannica) [reply src/zenml/zen_server/cache.py:65]: Ah, you're right, this was left behind from an older implementation version.

File: src/zenml/zen_server/cache.py
Comment (stefannica) [reply src/zenml/zen_server/cache.py:134]: I read that `OrderedDict` isn't available as a type annotation in Python 3.8 and decided to use `Dict` instead. But then I realized we don't support Python 3.8 anymore so... yeah, sure :)

File: src/zenml/zen_server/cache.py
Comment (stefannica) [reply src/zenml/zen_server/cache.py]: done

File: src/zenml/zen_server/cache.py
Comment (stefannica) [reply src/zenml/zen_server/cache.py]: done

File: src/zenml/zen_server/cache.py
Comment (stefannica) [reply src/zenml/zen_server/cache.py]: done

**Review by schustmi — APPROVED** [review]

**Pull request merged**
